processmanager/processinfo.go

```diff
@@ -720,29 +697,42 @@ func (pm *ProcessManager) MetaForPID(pid libpf.PID) ProcessMeta {
 
 // findMappingForTrace locates the mapping for a given host trace.
 func (pm *ProcessManager) findMappingForTrace(pid libpf.PID, fid host.FileID,
-	addr libpf.AddressOrLineno) (m Mapping, found bool) {
-	pm.mu.RLock()
-	defer pm.mu.RUnlock()
+	addr libpf.Address) libpf.FrameMapping {
+	var maps []Mapping
 
-	procInfo, ok := pm.pidToProcessInfo[pid]
-	if !ok {
-		return Mapping{}, false
+	pm.mu.RLock()
+	if procInfo, ok := pm.pidToProcessInfo[pid]; ok {
+		maps = procInfo.mappings
 	}
-
-	fidMappings, ok := procInfo.mappingsByFileID[fid]
-	if !ok {
-		return Mapping{}, false
+	pm.mu.RUnlock()
+	if maps == nil {
+		return libpf.FrameMapping{}
 	}
 
-	for _, candidate := range fidMappings {
-		procSpaceVA := libpf.Address(uint64(addr) + candidate.Bias)
-		mappingEnd := candidate.Vaddr + libpf.Address(candidate.Length)
-		if procSpaceVA >= candidate.Vaddr && procSpaceVA <= mappingEnd {
-			return *candidate, true
+	i := sort.Search(len(maps), func(i int) bool {
+		entry := &maps[i]
+		fm := entry.FrameMapping.Value()
+		f := fm.File.Value()
+		entryFid := host.FileIDFromLibpf(f.FileID)
+		if entryFid != fid {
+			return entryFid >= fid
```

Comment:
As FileIDs are not directly comparable, we should return false here.
```suggestion
			return false
		}
		return fm.Start >= addr
```